src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,204 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : Component, IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _transactionManagerLock = new object();
+        private readonly IThreadProvider _threadProvider;
+
+        public DistributedLockService(IMachineNameProvider machineNameProvider, IThreadProvider threadProvider, ILifetimeScope lifetimeScope, IClock clock) {
+            _machineNameProvider = machineNameProvider;
+            _lifetimeScope = lifetimeScope;
+            _clock = clock;
+            _threadProvider = threadProvider;
+        }
+
+        public bool TryAcquireLockForMachine(string name, TimeSpan maxValidFor, TimeSpan? timeout, out DistributedLock @lock) {
+            return TryAcquireLock(name, maxValidFor, timeout, GetMachineName(), null, out @lock);
+        }
+
+        public DistributedLock AcquireLockForMachine(string name, TimeSpan maxValidFor, TimeSpan? timeout) {
+            return AcquireLock(name, maxValidFor, timeout, GetMachineName(), null);
+        }
+
+        public bool TryAcquireLockForThread(string name, TimeSpan maxValidFor, TimeSpan? timeout, out DistributedLock @lock) {
+            return TryAcquireLock(name, maxValidFor, timeout, GetMachineName(), GetThreadId(), out @lock);
+        }
+
+        public DistributedLock AcquireLockForThread(string name, TimeSpan maxValidFor, TimeSpan? timeout) {
+            return AcquireLock(name, maxValidFor, timeout, GetMachineName(), GetThreadId());
+        }
+
+        public void ReleaseLock(DistributedLock @lock) {
+            lock (_transactionManagerLock) {
```

Comment:
Yea will check. In theory each new lifetime scope should create a new transaction manager, so we are not sure if we need the lock at all. Perhaps it's the way the services are registered in the testing environment. Will investigate (as discussed offline).